### PR TITLE
Prefer `jnp.tile` over `concatenate`

### PR DESCRIPTION
**Pull request by lgeiger**
`jnp.tile` only uses broadcasting which tends to be preferred so this PR replaces `concatenate` with `jnp.tile` in the cases where it is possible.

### CODE REVIEW

**Comment by YouJiacheng** [comment]
I think we can use `jax.lax.broadcast_in_dim` + `jax.lax.reshape` for faster tracing time. There are lots of python code in `jnp.tile` and `jnp.broadcast_to` (used by `jnp.tile`).

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -3245,7 +3245,9 @@ def _roll(a, shift, axis):
                   np.broadcast_to(axis, b_shape)):
     i = _canonicalize_axis(i, a_ndim)
     x = remainder(x, (a_shape[i] or 1))
-    a = lax.concatenate((a, a), i)
+    reps = [1] * a_ndim
+    reps[i] = 2
+    a = tile(a, reps)
```
[the remark this answers is not in the document]
Comment (YouJiacheng) [reply jax/_src/numpy/lax_numpy.py:3250]: Or maybe we can simplify implementation of `jnp.tile`?

**Comment by YouJiacheng** [comment]
Or, for simplicity and better semantic we should use `jax.numpy.resize` in the first case.

File: jax/_src/prng.py [hunk 2]
```diff
@@ -580,7 +580,7 @@ def threefry_random_bits(key: jnp.ndarray, bit_width, shape):
 
 def _rbg_seed(seed: int) -> jnp.ndarray:
   halfkey = threefry_seed(seed)
-  return jnp.concatenate([halfkey, halfkey])
+  return jnp.tile(halfkey, 2)
```
Comment (YouJiacheng) [inline jax/_src/prng.py:583]: concatenate repeat axis 0, while tile repeat axis - 1. Are you sure halfkey is 1d array?

Comment (lgeiger) [reply jax/_src/prng.py:583]: Yes, this seems to be the case:
https://github.com/google/jax/blob/3bfa6af2c81ac3c08b05d49b70c1bf01f09d28c5/jax/_src/prng.py#L246-L256

Comment (jakevdp) [reply jax/_src/prng.py:583]: I think this will probably cause issues with custom PRNG keys, which don't concatenate like regular arrays. (`jnp.concatenate` checks for this, while `jnp.tile` does not). I'm going to request a review from @froystig, who would probably have recommendations here.

Comment (froystig) [reply jax/_src/prng.py:583]: Good thinking @jakevdp. Actually, in this case, `tile` is valid use and is already supported! For more on how and why see #8381.

**Review by froystig — APPROVED** [review]

**Comment by froystig** [comment]
@lgeiger, just to make sure you noticed: the CI checks show that one unit test (`LaxBackedNumpyTests.testNonScalarRepeats_fixed_size=False`) fails on python 3.7 with our oldest supported version of numpy (1.19.5).

**Comment by lgeiger** [comment]
> just to make sure you noticed: the CI checks show that one unit test (`LaxBackedNumpyTests.testNonScalarRepeats_fixed_size=False`) fails on python 3.7 with our oldest supported version of numpy (1.19.5)

@froystig Sorry for the delay, but unfortunately I am unable to reproduce the test failure with Numpy 1.19.5 on macOS or even in a Python 3.7 Docker container. I am not sure what could cause this problem, since it doesn't seem to reproduce on all system.
For now I rebased my PR onto the latest main, maybe this will fix it? But this could also be indicative of some underlying problem that is platform specific.

**Comment by froystig** [comment]
Thanks for taking a look and rebasing. Seems fine now...
